### PR TITLE
[Merged by Bors] - feat: port Algebra.Category.GroupWithZero

### PR DESCRIPTION
**Pull request by pimotte**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/GroupWithZero.lean [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+instance hasForgetToMon : HasForget₂ GroupWithZeroCat MonCat
+    where forget₂ :=
+      { obj := fun X => ⟨ X , _ ⟩
+        map := fun f => f.toMonoidHom }
+#align GroupWithZero.has_forget_to_Mon GroupWithZeroCat.hasForgetToMon
+
+/-- Constructs an isomorphism of groups with zero from a group isomorphism between them. -/
+@[simps]
+def Iso.mk {α β : GroupWithZeroCat.{u}} (e : α ≃* β) : α ≅ β where
+  hom := e.toFun
```
Comment (pimotte) [inline Mathlib/Algebra/Category/GroupWithZero.lean:96]: This yields the following error:

```
type mismatch
  e.toFun
has type
  ↑α → ↑β : Type u
but is expected to have type
  α ⟶ β : Type u
  ```
  
  Without the toFun, it gives a similar error (with `α ≃* β` instead)

Comment (pimotte) [reply Mathlib/Algebra/Category/GroupWithZero.lean]: Fixed

**Comment by pimotte** [comment]
Should I rename the file to GroupWithZeroCat? The porting notes say to do so if it causes name clashes, but I don't seem to be running into anything problematic.

**Comment by semorrison** [comment]
> Should I rename the file to GroupWithZeroCat? The porting notes say to do so if it causes name clashes, but I don't seem to be running into anything problematic.
> 
> Main question for review: Are the instances in the right place?

Yes, I think we've realised that for now we need the `Cat` suffix sufficiently often that it is the default until we bikeshed this again.

File: Mathlib/Algebra/Category/GroupWithZero.lean [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+-- porting note: was not necessary in mathlib
+instance {M N : GroupWithZeroCat} : FunLike (M ⟶ N) M (fun _ => N) :=
+  ⟨ fun f => f.toFun, fun f g h => by
+    cases f
+    cases g
+    congr
+    apply FunLike.coe_injective'
+    exact h
+     ⟩
```
Comment (semorrison) [inline Mathlib/Algebra/Category/GroupWithZero.lean:64]:
```suggestion
  ⟨fun f => f.toFun, fun f g h => by
    cases f
    cases g
    congr
    apply FunLike.coe_injective'
    exact h⟩
```
Very minor, but we tend not to put whitespace around brackets.

File: Mathlib/Algebra/Category/GroupWithZero.lean [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+-- porting note: added
+lemma coeId {X : GroupWithZeroCat} : (𝟙 X : X → X) = id := rfl
+
+-- porting note: added
+lemma coeComp {X Y Z : GroupWithZeroCat} {f : X ⟶ Y} {g : Y ⟶ Z} : (f ≫ g : X → Z) = g ∘ f := rfl
+
+instance groupWithZeroConcreteCategory : ConcreteCategory GroupWithZeroCat where
+  forget := { obj := fun G => G
+              map := fun f => f.toFun }
+  forget_faithful := ⟨ fun h => FunLike.coe_injective h ⟩
```
Comment (semorrison) [inline Mathlib/Algebra/Category/GroupWithZero.lean:75]: Again, not essential, but I would format this as:
```suggestion
  forget :=
  { obj := fun G => G
    map := fun f => f.toFun }
  forget_faithful := ⟨fun h => FunLike.coe_injective h⟩
```

File: Mathlib/Algebra/Category/GroupWithZero.lean [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  ⟨ fun f => f.toFun, fun f g h => by
+    cases f
+    cases g
+    congr
+    apply FunLike.coe_injective'
+    exact h
+     ⟩
+
+-- porting note: added
+lemma coeId {X : GroupWithZeroCat} : (𝟙 X : X → X) = id := rfl
```
Comment (semorrison) [inline Mathlib/Algebra/Category/GroupWithZero.lean:67]: Lemma names should be snake cased, so this is `coe_id`.

File: Mathlib/Algebra/Category/GroupWithZero.lean [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+instance groupWithZeroConcreteCategory : ConcreteCategory GroupWithZeroCat where
+  forget := { obj := fun G => G
+              map := fun f => f.toFun }
+  forget_faithful := ⟨ fun h => FunLike.coe_injective h ⟩
+
+-- porting note: added
+@[simp] lemma forget_map (f : X ⟶ Y) : (forget GroupWithZeroCat).map f = f := rfl
+
+
```
Comment (semorrison) [inline Mathlib/Algebra/Category/GroupWithZero.lean:80]:
```suggestion

```

File: Mathlib/Algebra/Category/GroupWithZero.lean [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+instance hasForgetToBipointed : HasForget₂ GroupWithZeroCat Bipointed
+    where forget₂ :=
+      { obj := fun X => ⟨X, 0, 1⟩
+        map := fun f => ⟨f, f.map_zero', f.map_one'⟩
+        }
```
Comment (semorrison) [inline Mathlib/Algebra/Category/GroupWithZero.lean:85]:
```suggestion
        map := fun f => ⟨f, f.map_zero', f.map_one'⟩ }
```

File: Mathlib/Algebra/Category/GroupWithZero.lean [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+-- porting note: this instance was not necessary in mathlib
+instance {X Y : GroupWithZeroCat} : CoeFun (X ⟶ Y) fun _ => X → Y where
+  coe (f : X →*₀ Y) := f
+
+/-- Conversion from MulEquiv to MonoidWithZeroHom -/
+-- porting note : this function was not necessary in mathlib
+def toMonoidWithZeroHom {M N} [GroupWithZero M] [GroupWithZero N] (h : M ≃* N) : M →*₀ N :=
+  {
+    toFun := h.toFun
```
Comment (semorrison) [inline Mathlib/Algebra/Category/GroupWithZero.lean:104]:
```suggestion
  { toFun := h.toFun
```

File: Mathlib/Algebra/Category/GroupWithZero.lean [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    map_mul' := h.map_mul'
+    map_one' := h.map_one
+    map_zero' := by
+       rw [← mul_eq_zero_of_left rfl 1]
+       rw [h.map_mul' 0 1]
+       simp only [MulEquiv.toEquiv_eq_coe, Equiv.toFun_as_coe_apply, MulEquiv.coe_toEquiv,
+         map_zero, map_one, mul_one]
+   }
```
Comment (semorrison) [inline Mathlib/Algebra/Category/GroupWithZero.lean:112]:
```suggestion
    map_zero' := by
      rw [← mul_eq_zero_of_left rfl 1]
      rw [h.map_mul' 0 1]
      simp only [MulEquiv.toEquiv_eq_coe, Equiv.toFun_as_coe_apply, MulEquiv.coe_toEquiv,
         map_zero, map_one, mul_one] }
```

File: Mathlib/Algebra/Category/GroupWithZero.lean [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+set_option linter.uppercaseLean3 false in
+#align GroupWithZero.has_forget_to_Mon GroupWithZeroCat.hasForgetToMon
+
+-- porting note: this instance was not necessary in mathlib
+instance {X Y : GroupWithZeroCat} : CoeFun (X ⟶ Y) fun _ => X → Y where
+  coe (f : X →*₀ Y) := f
+
+/-- Conversion from MulEquiv to MonoidWithZeroHom -/
+-- porting note : this function was not necessary in mathlib
+def toMonoidWithZeroHom {M N} [GroupWithZero M] [GroupWithZero N] (h : M ≃* N) : M →*₀ N :=
```
Comment (semorrison) [inline Mathlib/Algebra/Category/GroupWithZero.lean:102]: I'm dubious adding this declaration is the right solution to the problem below. In Lean 3 this was being done by a coercion, right? We should make sure the same coercion works here. Perhaps in `Iso.mk` you can write a type ascription, e.g. `(e : M →*₀ N)`?

Please ping me on zulip if you need further help with this.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**